pkg/ebpf/c/tracee.bpf.c

```diff
@@ -2008,128 +2162,210 @@ static __always_inline int do_should_trace(program_data_t *p)
+        //                               0000 0010 |
+        //                               1111 0001
+        //                               ---------
+        //                               1111 0011 (scope 3 and 4 unmatched)
+        //
+        // considering an event from "ping" command
+        //
+        // res                         = 1111 1111 <- ~0ULL
+        // equality_filter_matches()   = 0000 0100
+        //
+        // res                         = 1111 1111 &
+        //                               0000 0100 |
+        //                               1111 0001
+        //                               ---------
+        //                               1111 0101 (scope 2 and 4 unmatched)
+
+        u64 filter_out_scopes = p->config->comm_filter_out_scopes;
+        u64 mask = ~p->config->comm_filter_enabled_scopes;
+        res &= equality_filter_matches(filter_out_scopes, &comm_filter, &context->comm) | mask;
+    }
+
+    if (p->config->proc_tree_filter_enabled_scopes) {
+        u64 filter_out_scopes = p->config->proc_tree_filter_out_scopes;
+        u64 mask = ~p->config->proc_tree_filter_enabled_scopes;
+        res &= equality_filter_matches(filter_out_scopes, &process_tree_map, &context->host_pid) |
+               mask;
+    }
+
+    if (p->config->cgroup_id_filter_enabled_scopes) {
+        u64 filter_out_scopes = p->config->cgroup_id_filter_out_scopes;
+        u64 mask = ~p->config->cgroup_id_filter_enabled_scopes;
+        u64 cgroup_id_lsb = context->cgroup_id;
+        res &= equality_filter_matches(filter_out_scopes, &cgroup_id_filter, &cgroup_id_lsb) | mask;
+    }
+
+    if (p->config->bin_path_filter_enabled_scopes) {
+        u64 filter_out_scopes = p->config->bin_path_filter_out_scopes;
+        u64 mask = ~p->config->bin_path_filter_enabled_scopes;
+        res &= binary_filter_matches(filter_out_scopes, proc_info) | mask;
+    }
+
+    if (p->config->follow_filter_enabled_scopes) {
+        // trace this proc anyway if follow was set by a scope
+        u64 mask = p->config->follow_filter_enabled_scopes;
+        res |= proc_info->follow_in_scopes & mask;
+    }
+
+    // e.g.: cmdline: -t 5:container
+    //
+    // considering an event from a container
+    //
+    // res = 1111 1111 &
+    //       0001 0000
+    //       ---------
+    //       0001 0000 (scope 5 match)
+    //
+    // considering an event not from a container
+    //
+    // res = 1110 1111 &
+    //       0001 0000
+    //       ---------
+    //       0000 0000 (no scope match)
+    //
+    // ---
+    //
+    // e.g.: cmdline: -t 2:comm=who -t 3:comm=ping -t 4:comm!=who
+    //
+    // considering an event from "who" command
+    //
+    // res = 1111 0011 &
+    //       0000 1110
+    //       ---------
+    //       0000 0010 (scope 2 match)
+    //
+    // considering an event from "ping" command
+    //
+    // res = 1111 0101 &
+    //       0000 1110
+    //       ---------
+    //       0000 0100 (scope 3 match)
```

Comment:
Again, I don't see a reason for repeating all of this.
The new thing in this line is ANDing with enabled_scopes so I would just explain about that in one sentence:
`Make sure only enabled scopes are set in the bitmask (other bits are invalid)`